### PR TITLE
Expression function `decode_uri`

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description
Adds a new expression function to get uri information from data providers of a layer. This can be used to determine the file path of ogr based layers.

`decode_uri(@layer, 'path')`

```yaml
name: decode_uri
type: function
description: Takes a layer and decodes the uri of the undelying data provider. It
  depends on the dataprovider, which data is available.
arguments:
- arg: layer
  description: The layer for which the uri should be decoded.
- arg: part
  description: The part of the uri to return. If unspecified, a map with all uri parts
    will be returned.
examples:
- expression: decode_uri(@layer)
  returns: "{'layerId': '0', 'layerName': '', 'path': '/home/qgis/shapefile.shp'}"
- expression: decode_uri(@layer)
  returns: "{'layerId': NULL, 'layerName': 'layer', 'path': '/home/qgis/geopackage.shp'}"
- expression: decode_uri(@layer), 'path'
  returns: "'/home/qgis/geopackage.shp'"

```

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice!

Instead of the second argument, have you thought about returning a map of all values of instead? I think that might be nicer to work with

**Comment by m-kuhn** [comment]
The second parameter is optional. If you leave it out, a map is returned.

>  Is-it possible to add the provider (postgis, wfs, ogr etc) in the dictionary?

It would be a nice addition too but I'd vote for putting it into a separate expression.

**Comment by m-kuhn** [comment]
@nirvn love it :heart: ! Can it also access arrays?
I'd still leave the second parameter optional, ok?

**Comment by nirvn** [comment]
@m-kuhn , yeah the goal is to support both map and array objects. Brainstorming with @nyalldawson on how best to achieve that.

I don't have a strong opinion on optional 2nd argument. I'd say I'm -0 mainly as a way to incite ppl to use square brackets syntax.

**Comment by nyalldawson** [comment]
Random thoughts I had overnight:

While those familiar with QGIS API (mostly? :stuck_out_tongue: ) understand the structure of layer sources with URIs, etc, I very much doubt that end users would have this association of a layer's file path with a function named "decode_uri".

I think if I was a user completely unaware of QGIS API, I'd expect to see the layer's path returned as part of the existing `layer_property` function.

Would you be against merging these two expression functions? (I realise that layer_property" doesn't currently supported the map type return, but that's been on my "to fix" for a while... layer_property predated the expression map support and needs to be modernised to utilise this too).

File: resources/function_help/json/decode_uri [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "decode_uri",
+  "type": "function",
+  "description": "Takes a layer and decodes the uri of the undelying data provider. It depends on the dataprovider, which data is available.",
+  "arguments": [ {"arg":"layer","description":"The layer for which the uri should be decoded."},
+                 {"arg":"part","description":"The part of the uri to return. If unspecified, a map with all uri parts will be returned."} ],
+  "examples": [ { "expression":"decode_uri(@layer)", "returns":"{'layerId': '0', 'layerName': '', 'path': '/home/qgis/shapefile.shp'}"},
+                { "expression":"decode_uri(@layer)", "returns":"{'layerId': NULL, 'layerName': 'layer', 'path': '/home/qgis/geopackage.shp'}"},
```
Comment (nyalldawson) [inline resources/function_help/json/decode_uri]: geopackage.shp ? :upside_down_face:

Comment (m-kuhn) [reply resources/function_help/json/decode_uri]: There's something in there for the fanboys and those who want stable, eh?

**Comment by m-kuhn** [comment]
> While those familiar with QGIS API (mostly? 😛 ) understand the structure of layer sources with URIs, etc, I very much doubt that end users would have this association of a layer's file path with a function named "decode_uri".
> I think if I was a user completely unaware of QGIS API, I'd expect to see the layer's path returned as part of the existing layer_property function.

Fair point. But the decode_uri is a flexible method which may return different things based on the provider whereas the layer_property list is a curated list of available keys.

* Merge the two
  * There's a risk of key collisions and it's hard to document what can be expected from layer_property
* Add 'path' (and other interesting parts) to layer_property too and leave decode_uri for those adventurous enough
  * Nice tradeoff?

File: resources/function_help/json/decode_uri [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "decode_uri",
+  "type": "function",
+  "description": "Takes a layer and decodes the uri of the underlying data provider. It depends on the dataprovider, which data is available.",
+  "arguments": [ {"arg":"layer","description":"The layer for which the uri should be decoded."},
+                 {"arg":"part","description":"The part of the uri to return. If unspecified, a map with all uri parts will be returned."} ],
+  "examples": [ { "expression":"decode_uri(@layer)", "returns":"{'layerId': '0', 'layerName': '', 'path': '/home/qgis/shapefile.shp'}"},
+                { "expression":"decode_uri(@layer)", "returns":"{'layerId': NULL, 'layerName': 'layer', 'path': '/home/qgis/geopackage.gpkg'}"},
+                { "expression":"decode_uri(@layer), 'path'", "returns":"'C:\my_data\qgis\shape.shp'"}
```
Comment (Gustry) [inline resources/function_help/json/decode_uri]: decode_uri(@layer, 'path')

File: resources/function_help/json/decode_uri [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "decode_uri",
+  "type": "function",
+  "description": "Takes a layer and decodes the uri of the underlying data provider. It depends on the dataprovider, which data is available.",
+  "arguments": [ {"arg":"layer","description":"The layer for which the uri should be decoded."},
+                 {"arg":"part","description":"The part of the uri to return. If unspecified, a map with all uri parts will be returned."} ],
```
Comment (Gustry) [inline resources/function_help/json/decode_uri]: Missing the optional flag

File: resources/function_help/json/decode_uri (hunk 2 above)
Comment (m-kuhn) [reply resources/function_help/json/decode_uri]: Ouch, lost in a force push after already fixed by @DelazJ

File: resources/function_help/json/decode_uri (hunk 3 above)
Comment (m-kuhn) [reply resources/function_help/json/decode_uri]: done, thx

**Pull request merged**